ivy_tests/test_ivy/test_functional/test_core/test_gradients.py

```diff
@@ -4,164 +4,196 @@
-
+@given(
+    dtype_and_x=helpers.dtype_and_values(ivy_np.valid_float_dtypes),
+    as_variable=st.booleans(),
+    native_array=st.booleans(),
+    num_positional_args=helpers.num_positional_args(fn_name="variable_data"),
+    container=st.booleans(),
+    instance_method=st.booleans(),
+)
+def test_variable_data(
+    dtype_and_x,
+    as_variable,
+    native_array,
+    num_positional_args,
+    container,
+    instance_method,
+    fw
+):
+    dtype, x = dtype_and_x
+    x = np.asarray(x, dtype=dtype)
+    if fw == "torch":
+        x = torch.as_tensor(x,dtype=torch.float32)
+    if fw == "tensorflow":
+        x = tf.Variable(tf.convert_to_tensor(x,dtype=dtype))
+    helpers.test_function(
+        dtype,
+        as_variable,
+        False,
+        native_array,
+        fw,
+        num_positional_args,
+        container,
+        instance_method,
+        "variable_data",
+        x=x
+    )
 
 # stop_gradient
-@pytest.mark.parametrize("x_raw", [[0.0]])
-@pytest.mark.parametrize("dtype", ["float32"])
-@pytest.mark.parametrize("tensor_fn", [("array", ivy.array), ("var", helpers.var_fn)])
-def test_stop_gradient(x_raw, dtype, tensor_fn, device, call):
-    # smoke test
-    fn_name, tensor_fn = tensor_fn
-    x = tensor_fn(x_raw, dtype=dtype, device=device)
-    ret = ivy.stop_gradient(x)
-    # type test
-    if fn_name == "array":
-        assert ivy.is_ivy_array(ret)
-    elif call is not helpers.np_call:
-        # Numpy does not support variables, is_variable() always returns False
-        assert ivy.is_variable(ret)
-    # cardinality test
-    assert ret.shape == x.shape
-    # value test
-    if call is not helpers.tf_graph_call:
-        # Tf graph mode cannot create variables as part of the computation graph
-        assert np.array_equal(
-            call(ivy.stop_gradient, x),
-            np.array(x_raw, dtype=dtype),
-        )
-    # compilation test
-    if call in [helpers.torch_call]:
-        # pytorch scripting does not support attribute setting
-        return
-
+@given(
+    dtype_and_x=helpers.dtype_and_values(ivy_np.valid_float_dtypes),
+    as_variable=st.booleans(),
+    native_array=st.booleans(),
+    num_positional_args=helpers.num_positional_args(fn_name="stop_gradient"),
+    container=st.booleans(),
+    instance_method=st.booleans(),
+    preserve_type=st.booleans()
+)
+def test_stop_gradient(
+    dtype_and_x,
+    as_variable,
+    native_array,
+    num_positional_args,
+    container,
+    instance_method,
+    fw,
+    preserve_type
+):
+    dtype, x = dtype_and_x
+    x = np.asarray(x, dtype=dtype)
+    helpers.test_function(
+        dtype,
+        as_variable,
+        False,
+        native_array,
+        fw,
+        num_positional_args,
+        container,
+        instance_method,
+        "stop_gradient",
+        x=x,
+        preserve_type=preserve_type
+    )
 
 # execute_with_gradients
-@pytest.mark.parametrize(
-    "func_n_xs_n_ty_n_te_n_tg",
-    [
-        (
+@given(
```

Comment:
this is okay for now, but once `helpers.test_function` supports self-consistency testing against an arbitrary backend framework, then this test should look similar to the other tests, with `torch` used as our reference backend framework. Until this is implemented, this hard-coded ground truth approach suffices.